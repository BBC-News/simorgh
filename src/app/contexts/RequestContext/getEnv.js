const getEnv = origin => {
  if (origin.includes('local')) {
    return 'local';
  }

  if (origin.includes('test')) {
    return 'test';
  }

  if (origin.includes('stage')) {
    return 'stage';
  }

  return 'live';
};

export default getEnv;
