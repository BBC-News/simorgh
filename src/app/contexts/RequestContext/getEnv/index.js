const getEnv = origin => {
  if (origin.includes('local')) {
    return 'local';
  }

  if (origin.includes('test')) {
    return 'test';
  }

  return 'live';
};

export default getEnv;
