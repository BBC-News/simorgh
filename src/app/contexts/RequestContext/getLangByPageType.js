import deepGet from '../../lib/utilities/deepGet';

const getLangByPageType = (data, defaultLang, pageType) => {
  return pageType === 'article'
    ? deepGet(['pageData', 'metadata', 'passport', 'language'], data)
    : defaultLang;
};

export default getLangByPageType;
