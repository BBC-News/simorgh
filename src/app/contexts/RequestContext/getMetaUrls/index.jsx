const getAmpUrl = url => `${url}.amp`;

const getCanonicalUrl = (origin, service, pageType, id) => {
  switch (pageType) {
    case 'frontPage':
      return `${origin}/${service}`;
    default:
      return `${origin}/${service}/articles/${id}`;
  }
};

const getCanonicalUrlWithVariant = (
  origin,
  service,
  serviceVariant,
  pageType,
  id,
) => {
  switch (pageType) {
    case 'frontPage':
      return `${origin}/${service}/${serviceVariant}`;
    default:
      return `${origin}/${service}/articles/${id}/${serviceVariant}`;
  }
};

const getUkCanonicalUrl = url => {
  const tldRegex = /(.com|.co.uk)/g;
  return url.replace(tldRegex, '.co.uk');
};

const getNonUkCanonicalUrl = url => {
  const tldRegex = /(.com|.co.uk)/g;
  return url.replace(tldRegex, '.com');
};

const getMetaUrls = (origin, service, serviceVariant, pageType, id) => {
  let canonicalLink = getCanonicalUrl(origin, service, pageType, id);
  let canonicalUkLink = getUkCanonicalUrl(canonicalLink);
  let canonicalNonUkLink = getNonUkCanonicalUrl(canonicalLink);

  if (serviceVariant) {
    canonicalLink = getCanonicalUrlWithVariant(
      origin,
      service,
      serviceVariant,
      pageType,
      id,
    );
    canonicalUkLink = getUkCanonicalUrl(canonicalLink);
    canonicalNonUkLink = getNonUkCanonicalUrl(canonicalLink);

    return {
      canonicalLink,
      ampLink: getAmpUrl(canonicalLink),
      canonicalUkLink,
      ampUkLink: getAmpUrl(canonicalUkLink),
      canonicalNonUkLink,
      ampNonUkLink: getAmpUrl(canonicalNonUkLink),
    };
  }

  return {
    canonicalLink,
    ampLink: getAmpUrl(canonicalLink),
    canonicalUkLink,
    ampUkLink: getAmpUrl(canonicalUkLink),
    canonicalNonUkLink,
    ampNonUkLink: getAmpUrl(canonicalNonUkLink),
  };
};

export default getMetaUrls;
