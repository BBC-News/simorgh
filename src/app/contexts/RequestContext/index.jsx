import React from 'react';
import { bool, node, oneOf, string } from 'prop-types';

export const RequestContext = React.createContext('default');

export const RequestContextProvider = ({
  children,
  env,
  id,
  isUK,
  lang,
  origin,
  pageType,
  platform,
  statsDestination,
  statsPageIdentifier,
}) => {
  const value = {
    env,
    id,
    isUK,
    lang,
    origin,
    pageType,
    platform,
    statsDestination,
    statsPageIdentifier,
  };

  return (
    <RequestContext.Provider value={value}>{children}</RequestContext.Provider>
  );
};

RequestContextProvider.propTypes = {
  children: node.isRequired,
  env: string,
  id: string,
  lang: string.isRequired,
  pageType: oneOf(['article', 'frontPage']).isRequired,
  platform: string.isRequired,
  isUK: bool.isRequired,
  origin: string.isRequired,
  statsDestination: string.isRequired,
  statsPageIdentifier: string.isRequired,
};

RequestContextProvider.defaultProps = {
  id: null,
  env: 'local',
};
