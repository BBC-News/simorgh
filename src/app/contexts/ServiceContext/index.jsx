import React from 'react';
import { node, string } from 'prop-types';
import services from '#lib/config/services/loadableConfig';
import variantPropType from '../../models/propTypes/variants';
import { getVariant } from '#lib/utilities/variantHandler';

export const ServiceContext = React.createContext({});

export const ServiceContextProvider = ({ children, service, variant }) => {
  const LoadableContextProvider = services[service];

  if (!LoadableContextProvider) {
    return null;
  }

  return (
    <LoadableContextProvider
      Context={ServiceContext}
      dataKey={getVariant({ service, variant })}
    >
      {children}
    </LoadableContextProvider>
  );
};

ServiceContextProvider.propTypes = {
  children: node.isRequired,
  service: string,
  variant: variantPropType,
};

ServiceContextProvider.defaultProps = {
  service: 'default',
  variant: 'default',
};
