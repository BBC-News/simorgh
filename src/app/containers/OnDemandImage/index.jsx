import React, { useContext } from 'react';
import { string } from 'prop-types';
import styled from 'styled-components';
import {
  GEL_SPACING_QUAD,
  GEL_SPACING_DBL,
  GEL_SPACING_TRPL,
} from '@bbc/gel-foundations/spacings';
import { GEL_GROUP_1_SCREEN_WIDTH_MAX } from '@bbc/gel-foundations/breakpoints';
import { ServiceContext } from '#contexts/ServiceContext';
import ImageWithPlaceholder from '../ImageWithPlaceholder';

const paddingDir = ({ dir }) => `padding-${dir === 'rtl' ? 'left' : 'right'}`;

const ImageContainer = styled.div`
  padding: ${GEL_SPACING_QUAD} 0 ${GEL_SPACING_TRPL};
  ${paddingDir}: ${GEL_SPACING_DBL};
  @media (max-width: ${GEL_GROUP_1_SCREEN_WIDTH_MAX}) {
    display: none;
  }
`;

const getSrc = ({ imageUrl, size }) =>
  `https://${imageUrl.replace('$recipe', `${size}x${size}`)}`;

const getSrcSet = ({ imageUrl, sizes }) =>
  sizes.map(size => `${getSrc({ imageUrl, size })} ${size}w`).join(',');

const smallImageSize = 128;
const mediumImageSize = 240;
const largeImageSize = 480;

const OnDemandImage = ({ imageUrl, dir }) => {
  const { defaultImageAltText: alt } = useContext(ServiceContext);

  const src = getSrc({ imageUrl, size: smallImageSize });
  const srcset = getSrcSet({
    imageUrl,
    sizes: [smallImageSize, mediumImageSize, largeImageSize],
  });
  const sizes = '(min-width: 1008px) 228px, 30vw';

  return (
    <ImageContainer dir={dir}>
      <ImageWithPlaceholder
        src={src}
        alt={alt}
        srcset={srcset}
        sizes={sizes}
        ratio={100}
        width={mediumImageSize}
        height={mediumImageSize}
      />
    </ImageContainer>
  );
};

OnDemandImage.propTypes = {
  imageUrl: string.isRequired,
  dir: string,
};

OnDemandImage.defaultProps = {
  dir: 'ltr',
};

export default OnDemandImage;
