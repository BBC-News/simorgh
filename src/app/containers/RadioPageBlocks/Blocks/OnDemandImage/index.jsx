import React, { useContext } from 'react';
import { string } from 'prop-types';
import styled from 'styled-components';
import Image, { AmpImg } from '@bbc/psammead-image';
import {
  GEL_SPACING_QUAD,
  GEL_SPACING_DBL,
  GEL_SPACING_TRPL,
} from '@bbc/gel-foundations/spacings';
import { GEL_GROUP_1_SCREEN_WIDTH_MAX } from '@bbc/gel-foundations/breakpoints';
import { RequestContext } from '#contexts/RequestContext';
import { ServiceContext } from '#contexts/ServiceContext';

const paddingDir = ({ dir }) => `padding-${dir === 'rtl' ? 'left' : 'right'}`;

const ImageContainer = styled.div`
  padding: ${GEL_SPACING_QUAD} 0 ${GEL_SPACING_TRPL};
  ${paddingDir}: ${GEL_SPACING_DBL};
  @media (max-width: ${GEL_GROUP_1_SCREEN_WIDTH_MAX}) {
    display: none;
  }
`;

const getSrc = ({ imageUrl, size }) =>
  `https://${imageUrl.replace('$recipe', `${size}x${size}`)}`;

const getSrcSet = ({ imageUrl, sizes }) =>
  sizes.map(size => `${getSrc({ imageUrl, size })} ${size}w`).join(',');

const smallImageSize = 128;
const mediumImageSize = 240;
const largeImageSize = 480;

const OnDemandImage = ({ imageUrl, dir }) => {
  const { isAmp } = useContext(RequestContext);
  const { defaultImageAltText: alt } = useContext(ServiceContext);

  const src = getSrc({ imageUrl, size: smallImageSize });
  const srcset = getSrcSet({
    imageUrl,
    sizes: [smallImageSize, mediumImageSize, largeImageSize],
  });
  const sizes = '(min-width: 1008px) 228px, 30vw';
  const imageProps = { src, alt, srcset };

  return (
    <ImageContainer dir={dir}>
      {isAmp ? (
        <AmpImg
          {...imageProps}
          layout="responsive"
          height={mediumImageSize}
          width={mediumImageSize}
        />
      ) : (
        <Image {...imageProps} sizes={sizes} />
      )}
    </ImageContainer>
  );
};

OnDemandImage.propTypes = {
  imageUrl: string.isRequired,
  dir: string,
};

OnDemandImage.defaultProps = {
  dir: 'ltr',
};

export default OnDemandImage;
