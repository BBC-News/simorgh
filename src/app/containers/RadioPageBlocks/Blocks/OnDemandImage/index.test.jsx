import React from 'react';
import { shouldMatchSnapshot } from '@bbc/psammead-test-helpers';
import { render, getByAltText } from '@testing-library/react';
import { ServiceContextProvider } from '#contexts/ServiceContext';
import { RequestContextProvider } from '#contexts/RequestContext';
import OnDemandImage from '.';

// eslint-disable-next-line react/prop-types
const component = ({ url, isAmp, service }) => (
  <ServiceContextProvider service={service}>
    <RequestContextProvider
      isAmp={isAmp}
      service={service}
      pageType="media"
      pathname="/path"
    >
      <OnDemandImage imageUrl={url} />
    </RequestContextProvider>
  </ServiceContextProvider>
);

describe('AudioPlayer blocks OnDemandHeading', () => {
  shouldMatchSnapshot(
    'should render correctly',
    component({ url: 'mock-url', isAmp: false, service: 'news' }),
  );

  it('should ensure the image has the right attributes', () => {
    const { container } = render(
      component({
        url: 'ichef.bbci.co.uk/images/ic/$recipe/p063j1dv.jpg',
        isAmp: false,
        service: 'pashto',
      }),
    );
    const img = getByAltText(container, 'BBC News پښتو');
    expect(img.src).toEqual(
      'https://ichef.bbci.co.uk/images/ic/128x128/p063j1dv.jpg',
    );
    expect(img.alt).toEqual('BBC News پښتو');
    expect(img.srcset).toEqual(
      'https://ichef.bbci.co.uk/images/ic/128x128/p063j1dv.jpg 128w,https://ichef.bbci.co.uk/images/ic/240x240/p063j1dv.jpg 240w,https://ichef.bbci.co.uk/images/ic/480x480/p063j1dv.jpg 480w',
    );
    expect(img.sizes).toEqual('(min-width: 1008px) 228px, 30vw');
  });

  it('should ensure the image has the right attributes for amp', () => {
    const { container } = render(
      component({
        url: 'ichef.bbci.co.uk/images/ic/$recipe/p063j1dv.jpg',
        isAmp: true,
        service: 'afaanoromoo',
      }),
    );
    const img = container.querySelector('amp-img');
    expect(img.getAttribute('src')).toEqual(
      'https://ichef.bbci.co.uk/images/ic/128x128/p063j1dv.jpg',
    );
    expect(img.getAttribute('alt')).toEqual('BBC News Afaan Oromoo');
    expect(img.getAttribute('layout')).toEqual('responsive');
    expect(img.getAttribute('srcset')).toEqual(
      'https://ichef.bbci.co.uk/images/ic/128x128/p063j1dv.jpg 128w,https://ichef.bbci.co.uk/images/ic/240x240/p063j1dv.jpg 240w,https://ichef.bbci.co.uk/images/ic/480x480/p063j1dv.jpg 480w',
    );
  });
});
