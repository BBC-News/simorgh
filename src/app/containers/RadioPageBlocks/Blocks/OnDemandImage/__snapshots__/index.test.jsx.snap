// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AudioPlayer blocks OnDemandHeading should render correctly 1`] = `
.c1 {
  display: block;
  width: 100%;
  visibility: visible;
}

.c0 {
  padding: 2rem 0 1.5rem;
  padding-right: 1rem;
}

@media (max-width:24.9375rem) {
  .c0 {
    display: none;
  }
}

<div
  class="c0"
  dir="ltr"
>
  <img
    alt="BBC News"
    class="c1"
    sizes="(min-width: 1008px) 228px, 30vw"
    src="https://mock-url"
    srcset="https://mock-url 128w,https://mock-url 240w,https://mock-url 480w"
  />
</div>
`;
